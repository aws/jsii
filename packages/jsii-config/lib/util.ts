import { readFile } from 'fs';

/*
 * Look for existing nested values in config, return undefined if not found
 */
export function getNestedValue(
  keys: string[],
  current: Record<string, unknown>,
): any {
  try {
    return keys.reduce((val: any, key: string) => val[key], current);
  } catch (_err) {
    return undefined;
  }
}

/*
 * recursively remove keys from object whose value empty.
 */
export function removeEmptyValues<T>(fields: T): T {
  return Object.entries(fields).reduce(
    (accum: any, [key, val]: [string, any]) => {
      if (typeof val === 'object') {
        return {
          ...accum,
          [key]: removeEmptyValues(val),
        };
      }

      return {
        ...accum,
        ...(!val
          ? {}
          : {
              [key]: val,
            }),
      };
    },
    {},
  );
}

/*
 * recursively flatten nested object
 */
export function flattenKeys(parent: string, fields: any) {
  return Object.entries(fields).reduce(
    (accum: any, [key, vals]: [string, any]) => ({
      ...accum,
      [`${parent}.${key}`]: vals,
    }),
    {},
  );
}

/**
 * utility for for fs.readFile as a promise
 *
 * TODO: use util.promisify without breaking test stub
 */
export async function readFilePromise(path: string): Promise<Buffer> {
  const result: Buffer = await new Promise((resolve, reject) => {
    readFile(path, (err, data) => {
      if (err) {
        return reject(err);
      }
      return resolve(data);
    });
  });
  return result;
}
