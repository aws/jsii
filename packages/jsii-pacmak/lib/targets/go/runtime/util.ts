import { CodeMaker } from 'codemaker';

import { JSII_INIT_ALIAS, JSII_INIT_FUNC } from './constants';

// Emits call to initialize runtime client if not already
export function emitInitialization(code: CodeMaker) {
  code.line(`${JSII_INIT_ALIAS}.${JSII_INIT_FUNC}()`);
}

/**
 * Slugify a name by appending '_' at the end until the resulting name is not
 * present in the list of reserved names.
 *
 * @param name     the name to be slugified
 * @param reserved the list of names that are already sued in-scope
 *
 * @returns the slugified name
 */
export function slugify(name: string, reserved: Iterable<string>): string {
  const used = new Set(reserved);
  while (used.has(name)) {
    name += '_';
  }
  return name;
}
