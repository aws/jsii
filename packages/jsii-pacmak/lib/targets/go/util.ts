import { Package } from './package';
import { GoMethod, GoTypeMember, GoType } from './types';

/*
 * Recursively search module for type with fqn
 */
export function findTypeInTree(
  module: Package,
  fqn: string,
): GoType | undefined {
  const result = module.types.find((t) => t.type.fqn === fqn);

  if (result) {
    return result;
  }

  return module.submodules.reduce((accum: GoType | undefined, sm) => {
    return accum || findTypeInTree(sm, fqn);
  }, undefined);
}

/*
 * Format NPM package names as idiomatic Go module name
 */
export function goPackageName(name: string): string {
  return name.replace(/[^a-z0-9.]/gi, '').toLowerCase();
}

export function flatMap<T, R>(
  collection: readonly T[],
  mapper: (value: T) => readonly R[],
): readonly R[] {
  return collection
    .map(mapper)
    .reduce((acc, elt) => acc.concat(elt), new Array<R>());
}

/*
 * Return module dependencies of a class or interface members
 */
export function getMemberDependencies(members: GoTypeMember[]): Package[] {
  return members.reduce((accum: Package[], member) => {
    return member.reference?.type?.pkg
      ? [...accum, member.reference?.type.pkg]
      : accum;
  }, []);
}

export function getParamDependencies(methods: GoMethod[]): Package[] {
  const dependencies: Package[] = [];
  for (const method of methods) {
    for (const param of method.parameters) {
      const pkg = param.reference?.type?.pkg;
      if (pkg) {
        dependencies.push(pkg);
      }
    }
  }
  return dependencies;
}

const RESERVED_WORDS: { [word: string]: string } = {
  break: 'break_',
  default: 'default_',
  func: 'func_',
  interface: 'interface_',
  select: 'select_',
  case: 'case_',
  defer: 'defer_',
  go: 'go_',
  map: 'map_',
  struct: 'struct_',
  chan: 'chan_',
  else: 'else_',
  goto: 'goto_',
  package: 'package_',
  switch: 'switch_',
  const: 'const_',
  fallthrough: 'fallthrough_',
  if: 'if_',
  range: 'range_',
  type: 'type_',
  continue: 'continue_',
  for: 'for_',
  import: 'import_',
  return: 'return_',
  var: 'var_',
};
/*
 * Sanitize reserved words
 */
export function substituteReservedWords(name: string): string {
  return RESERVED_WORDS[name] || name;
}
