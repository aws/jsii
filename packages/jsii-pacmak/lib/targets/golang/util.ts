import { Package, ModuleType } from './package';
import { TypeField } from './types';

/*
 * Recursively search module for type with fqn
 */
export function findTypeInTree(
  module: Package,
  fqn: string,
): ModuleType | undefined {
  const result = module.types.find((t) => t.type.fqn === fqn);

  if (result) {
    return result;
  }

  return module.submodules.reduce((accum: ModuleType | undefined, sm) => {
    return accum || findTypeInTree(sm, fqn);
  }, undefined);
}

/*
 * Format NPM package names as idiomatic Go module name
 */
export function goPackageName(name: string): string {
  return name.replace(/[^a-z0-9.]/gi, '').toLowerCase();
}

export function flatMap<T, R>(
  collection: readonly T[],
  mapper: (value: T) => readonly R[],
): readonly R[] {
  return collection
    .map(mapper)
    .reduce((acc, elt) => acc.concat(elt), new Array<R>());
}

/*
 * Return module dependencies of a class or interface fields
 */
export function getFieldDependencies(fields: TypeField[]): Package[] {
  return fields.reduce((accum: Package[], field) => {
    return field.references?.type?.parent
      ? [...accum, field.references?.type.parent]
      : accum;
  }, []);
}

const RESERVED_WORDS: { [word: string]: string } = {
  map: 'map_',
};
/*
 * Sanitize reserved words
 */
export function substituteReservedWords(name: string): string {
  return RESERVED_WORDS[name] || name;
}
