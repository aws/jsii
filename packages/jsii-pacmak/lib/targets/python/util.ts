const PYTHON_KEYWORDS = new Set([
  'False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class',
  'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from',
  'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass',
  'raise', 'return', 'try', 'while', 'with', 'yield'
]);

export function die(message: string): never {
  throw new Error(message);
}

export function toPythonIdentifier(name: string): string {
  if (PYTHON_KEYWORDS.has(name)) {
    return `${name}_`;
  }
  return name;
}
