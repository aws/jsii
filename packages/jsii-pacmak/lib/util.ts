import { spawn, SpawnOptions } from 'child_process';
import * as fs from 'fs-extra';
import * as spec from '@jsii/spec';
import * as os from 'os';
import * as path from 'path';
import * as logging from './logging';

export interface ShellOptions extends SpawnOptions {
  /**
   * Retry execution up to 3 times if it fails
   *
   * @default false
   */
  retry?: boolean;
}

/**
 * Given an npm package directory and a dependency name, returns the package directory of the dep.
 * @param packageDir     the root of the package declaring the dependency.
 * @param dependencyName the name of the dependency to be resolved.
 * @return the resolved directory path.
 */
export function resolveDependencyDirectory(
  packageDir: string,
  dependencyName: string,
): string {
  const lookupPaths = [path.join(packageDir, 'node_modules')];
  return path.dirname(
    require.resolve(`${dependencyName}/package.json`, { paths: lookupPaths }),
  );
}

export async function shell(
  cmd: string,
  args: string[],
  options: ShellOptions = {},
): Promise<string> {
  // eslint-disable-next-line @typescript-eslint/require-await
  async function spawn1() {
    logging.debug(cmd, args.join(' '), JSON.stringify(options));
    return new Promise<string>((ok, ko) => {
      const child = spawn(cmd, args, {
        ...options,
        shell: true,
        env: { ...process.env, ...(options.env ?? {}) },
        stdio: ['ignore', 'pipe', 'pipe'],
      });
      const stdout = new Array<Buffer>();
      const stderr = new Array<Buffer>();
      child.stdout.on('data', (chunk) => {
        if (logging.level >= logging.LEVEL_VERBOSE) {
          process.stderr.write(chunk); // notice - we emit all build output to stderr
        }
        stdout.push(Buffer.from(chunk));
      });
      child.stderr.on('data', (chunk) => {
        if (logging.level >= logging.LEVEL_VERBOSE) {
          process.stderr.write(chunk);
        }
        stderr.push(Buffer.from(chunk));
      });
      child.once('error', ko);

      // Must use CLOSE instead of EXIT; EXIT may fire while there is still data in the
      // I/O pipes, which we will miss if we return at that point.
      child.once('close', (code, signal) => {
        const out = Buffer.concat(stdout).toString('utf-8');
        if (code === 0) {
          return ok(out);
        }
        const err = Buffer.concat(stderr).toString('utf-8');
        const reason = signal != null ? `signal ${signal}` : `status ${code}`;
        const command = `${cmd} ${args.join(' ')}`;
        return ko(
          new Error(
            [
              `Command (${command}) failed with ${reason}:`,
              // STDERR first, the erro message could be truncated in logs.
              prefix(err, '#STDERR> '),
              prefix(out, '#STDOUT> '),
            ].join('\n'),
          ),
        );

        function prefix(text: string, add: string): string {
          return text
            .split('\n')
            .map((line) => `${add}${line}`)
            .join('\n');
        }
      });
    });
  }
  /* eslint-enable @typescript-eslint/require-await */

  let attempts = options.retry ? 3 : 1;
  while (attempts > 0) {
    attempts--;
    try {
      return spawn1();
    } catch (e) {
      if (attempts === 0) {
        throw e;
      }
      logging.info(`${e.message} (retrying)`);
    }
  }
  throw new Error('No attempts left'); // This is, in fact, unreachable code.
}

/**
 * Loads the assembly from a given module root directory.
 *
 * @param modulePath the path at which the node module is located.
 *
 * @return the parsed ``Assembly``.
 *
 * @throws if the module does not contain a JSII assembly file, or if it's invalid.
 */
export async function loadAssembly(modulePath: string): Promise<spec.Assembly> {
  const assmPath = path.join(modulePath, spec.SPEC_FILE_NAME);
  if (!(await fs.pathExists(assmPath))) {
    throw new Error(`Could not find ${assmPath}. Was the module built?`);
  }
  return spec.validateAssembly(await fs.readJson(assmPath));
}

/**
 * Strip filesystem unsafe characters from a string
 */
export function slugify(x: string) {
  return x.replace(/[^a-zA-Z0-9_-]/g, '_');
}

/**
 * Class that makes a temporary directory and holds on to an operation object
 */
export class Scratch<A> {
  public static async make<A>(
    factory: (dir: string) => Promise<A>,
  ): Promise<Scratch<A>>;
  public static async make<A>(factory: (dir: string) => A): Promise<Scratch<A>>;
  public static async make<A>(factory: (dir: string) => A | Promise<A>) {
    const tmpdir = await fs.mkdtemp(path.join(os.tmpdir(), 'npm-pack'));
    return new Scratch(tmpdir, await factory(tmpdir), false);
  }

  public static fake<A>(directory: string, object: A) {
    return new Scratch(directory, object, true);
  }

  public static async cleanupAll<A>(tempDirs: Array<Scratch<A>>) {
    await Promise.all(tempDirs.map((t) => t.cleanup()));
  }

  private constructor(
    public readonly directory: string,
    public readonly object: A,
    private readonly fake: boolean,
  ) {}

  public async cleanup() {
    if (!this.fake) {
      await fs.remove(this.directory);
    }
  }
}

export function setExtend<A>(xs: Set<A>, els: Iterable<A>) {
  for (const el of els) {
    xs.add(el);
  }
}

export async function filterAsync<A>(
  xs: A[],
  pred: (x: A) => Promise<boolean>,
): Promise<A[]> {
  const mapped = await Promise.all(
    xs.map(async (x) => ({ x, pred: await pred(x) })),
  );
  return mapped.filter(({ pred }) => pred).map(({ x }) => x);
}
