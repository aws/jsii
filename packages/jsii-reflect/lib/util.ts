
export function indexBy<T>(xs: T[], f: (x: T) => string): {[key: string]: T} {
  const ret: {[key: string]: T} = {};
  for (const x of xs) {
    ret[f(x)] = x;
  }
  return ret;
}