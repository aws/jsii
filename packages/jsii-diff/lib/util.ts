import * as childProcess from 'child_process';
import * as fs from 'fs-extra';
import * as log4js from 'log4js';
import * as os from 'os';
import * as path from 'path';
import * as util from 'util';

const LOG = log4js.getLogger('jsii-diff');

const exec = util.promisify(childProcess.exec);

export async function inTempDir<T>(block: () => T | Promise<T>): Promise<T> {
  const origDir = process.cwd();
  const tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'jsii'));
  process.chdir(tmpDir);
  try {
    return await block();
  } finally {
    process.chdir(origDir);
    await fs.remove(tmpDir);
  }
}

export type DownloadFailure = 'no_such_package';

export type NpmDownloadResult<T> =
  | { success: true; result: T }
  | { success: false; reason: DownloadFailure };

export function showDownloadFailure(f: DownloadFailure) {
  switch (f) {
    case 'no_such_package':
      return 'NPM package does not exist';
    default:
      return undefined;
  }
}

export async function downloadNpmPackage<T>(
  pkg: string,
  block: (dir: string) => Promise<T>,
): Promise<NpmDownloadResult<T>> {
  return inTempDir(async () => {
    LOG.info(`Fetching NPM package ${pkg}`);

    try {
      // Need to install package and dependencies in order for jsii-reflect
      // to not bork when it can find the dependencies.
      await exec(`npm install --silent --prefix . ${pkg}`);
    } catch (e) {
      // If this fails, might be because the package doesn't exist
      if (!isSubprocesFailedError(e)) {
        throw e;
      }
      if (await npmPackageExists(pkg)) {
        throw new Error(`NPM fetch failed: ${e}. Please try again.`);
      }
      LOG.warn(`NPM package ${pkg} does not exist.`);
      return { success: false, reason: 'no_such_package' } as NpmDownloadResult<
        T
      >;
    }

    const pkgDir = trimVersionString(pkg);
    return {
      success: true,
      result: await block(path.join(process.cwd(), 'node_modules', pkgDir)),
    } as NpmDownloadResult<T>;
  });
}

function isSubprocesFailedError(e: any) {
  return e.code !== undefined && e.cmd !== undefined;
}

async function npmPackageExists(pkg: string): Promise<boolean> {
  try {
    LOG.info(`Checking existence of ${pkg}`);
    await exec(`npm show --silent ${pkg}`);
    return true;
  } catch (e) {
    if (!isSubprocesFailedError(e)) {
      throw e;
    }
    return false;
  }
}

/**
 * Trim an optional version string from an NPM package name
 */
function trimVersionString(pkg: string) {
  // The arbitrary char before the @ prevents matching a @ at the start of the
  // string.
  return pkg.replace(/(.)@.*$/, '$1');
}

export function flatMap<T, U>(xs: T[], fn: (x: T) => U[]): U[] {
  const ret = new Array<U>();
  for (const x of xs) {
    ret.push(...fn(x));
  }
  return ret;
}

/**
 * Don't recurse infinitely by guarding a block with `do()`.
 */
export class RecursionBreaker<A> {
  private readonly elements = new Set<A>();

  public do(key: A, block: () => void) {
    if (this.elements.has(key)) {
      return;
    }

    this.elements.add(key);
    try {
      block();
    } finally {
      this.elements.delete(key);
    }
  }
}
