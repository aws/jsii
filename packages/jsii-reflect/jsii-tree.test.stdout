assemblies
 ├── jsii-calc
 ├── @scope/jsii-calc-base
 ├── @scope/jsii-calc-base-of-base
 └── @scope/jsii-calc-lib
