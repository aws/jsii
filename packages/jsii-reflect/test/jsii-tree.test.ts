import * as childProcess from 'child_process';
import * as path from 'path';
import { promisify } from 'util';

const exec = promisify(childProcess.exec);

test('jsii-tree', () => expect(jsiiTree()).resolves.toMatchSnapshot());

test('jsii-tree --all', () =>
  expect(jsiiTree('--all')).resolves.toMatchSnapshot());

test('jsii-tree --types', () =>
  expect(jsiiTree('--types')).resolves.toMatchSnapshot());

test('jsii-tree --members', () =>
  expect(jsiiTree('--members')).resolves.toMatchSnapshot());

test('jsii-tree --inheritance', () =>
  expect(jsiiTree('--inheritance')).resolves.toMatchSnapshot());

test('jsii-tree --signatures', () =>
  expect(jsiiTree('--signatures')).resolves.toMatchSnapshot());

async function jsiiTree(...args: string[]) {
  const command = [
    process.execPath,
    ...process.execArgv,
    path.join(__dirname, '..', 'bin', 'jsii-tree'),
    args.join(' '),
    '--no-colors',
    path.dirname(require.resolve('jsii-calc/package.json')),
  ].join(' ');

  const { stdout, stderr } = await exec(command);

  if (stderr) {
    console.error(stderr);
  }

  return stdout;
}
