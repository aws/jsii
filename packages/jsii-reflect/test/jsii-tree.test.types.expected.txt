assemblies
 ├─┬ jsii-calc
 │ └─┬ types
 │   ├── class AbstractClass
 │   ├── class AbstractClassBase
 │   ├── class AbstractClassReturner
 │   ├── class Add
 │   ├── class AllTypes
 │   ├── class AllowedMethodNames
 │   ├── class AsyncVirtualMethods
 │   ├── class AugmentableClass
 │   ├── class BinaryOperation
 │   ├── class Calculator
 │   ├── class ClassThatImplementsTheInternalInterface
 │   ├── class ClassThatImplementsThePrivateInterface
 │   ├── class ClassWithMutableObjectLiteralProperty
 │   ├── class ClassWithPrivateConstructorAndAutomaticProperties
 │   ├── class Constructors
 │   ├── class ConsumersOfThisCrazyTypeSystem
 │   ├── class DefaultedConstructorArgument
 │   ├── class Base
 │   ├── class Derived
 │   ├── class DoNotOverridePrivates
 │   ├── class DoNotRecognizeAnyAsOptional
 │   ├── class DontComplainAboutVariadicAfterOptional
 │   ├── class DoubleTrouble
 │   ├── class EraseUndefinedHashValues
 │   ├── class ExportedBaseClass
 │   ├── class GiveMeStructs
 │   ├── class GreetingAugmenter
 │   ├── class ImplementInternalInterface
 │   ├── class ImplementsInterfaceWithInternal
 │   ├── class ImplementsInterfaceWithInternalSubclass
 │   ├── class ImplementsPrivateInterface
 │   ├── class InbetweenClass
 │   ├── class Foo
 │   ├── class JSObjectLiteralForInterface
 │   ├── class JSObjectLiteralToNative
 │   ├── class JSObjectLiteralToNativeClass
 │   ├── class JavaReservedWords
 │   ├── class JsiiAgent
 │   ├── class Multiply
 │   ├── class Negate
 │   ├── class NodeStandardLibrary
 │   ├── class NullShouldBeTreatedAsUndefined
 │   ├── class NumberGenerator
 │   ├── class ObjectRefsInCollections
 │   ├── class OptionalConstructorArgument
 │   ├── class OverrideReturnsObject
 │   ├── class Polymorphism
 │   ├── class Power
 │   ├── class PublicClass
 │   ├── class PythonReservedWords
 │   ├── class ReferenceEnumFromScopedPackage
 │   ├── class ReturnsPrivateImplementationOfInterface
 │   ├── class RuntimeTypeChecking
 │   ├── class Statics
 │   ├── class StripInternal
 │   ├── class Sum
 │   ├── class SyncVirtualMethods
 │   ├── class Thrower
 │   ├── class UnaryOperation
 │   ├── class UseBundledDependency
 │   ├── class UseCalcBase
 │   ├── class UsesInterfaceWithProperties
 │   ├── class VariadicMethod
 │   ├── class VirtualMethodPlayground
 │   ├── class CompositeOperation
 │   ├── interface CalculatorProps
 │   ├── interface DerivedStruct
 │   ├── interface EraseUndefinedHashValuesOptions
 │   ├── interface ExtendsInternalInterface
 │   ├── interface ExtendsPrivateInterface
 │   ├── interface IAnotherPublicInterface
 │   ├── interface IFriendlier
 │   ├── interface IFriendlyRandomGenerator
 │   ├── interface IInterfaceThatShouldNotBeADataType
 │   ├── interface IInterfaceWithInternal
 │   ├── interface IInterfaceWithMethods
 │   ├── interface IInterfaceWithOptionalMethodArguments
 │   ├── interface IInterfaceWithProperties
 │   ├── interface IInterfaceWithPropertiesExtension
 │   ├── interface IMutableObjectLiteral
 │   ├── interface INonInternalInterface
 │   ├── interface IPrivatelyImplemented
 │   ├── interface IPublicInterface
 │   ├── interface IRandomNumberGenerator
 │   ├── interface IReturnsNumber
 │   ├── interface ImplictBaseOfBase
 │   ├── interface InterfaceImplementedByAbstractClass
 │   ├── interface Hello
 │   ├── interface Hello
 │   ├── interface LoadBalancedFargateServiceProps
 │   ├── interface NullShouldBeTreatedAsUndefinedData
 │   ├── interface UnionProperties
 │   ├── enum AllTypesEnum
 │   ├── enum StringEnum
 │   └── enum CompositionStringStyle
 ├─┬ @scope/jsii-calc-base
 │ └─┬ types
 │   ├── class Base
 │   └── interface BaseProps
 ├─┬ @scope/jsii-calc-base-of-base
 │ └─┬ types
 │   ├── class Very
 │   └── interface VeryBaseProps
 └─┬ @scope/jsii-calc-lib
   └─┬ types
     ├── class Number
     ├── class Operation
     ├── class Value
     ├── interface IDoublable
     ├── interface IFriendly
     ├── interface MyFirstStruct
     ├── interface StructWithOnlyOptionals
     └── enum EnumFromScopedModule
